gymnasium_robotics/envs/multiagent_mujoco/mujoco_multi.py

```diff
@@ -0,0 +1,636 @@
+
+    def map_local_actions_to_global_action(
+        self, actions: dict[str, numpy.array]
+    ) -> numpy.array:
+        """
+        Maps actions back into MuJoCo action space
+        Args:
+            action: An dict representing the action of each agent
+        Returns:
+            The action of the whole domain (is what eqivilent single agent action would be)
+        Raises:
+            AssertionError:
+                If the Agent action factorization is badly defined (if an action is double defined or not defined at all)
+        """
+        if self.agent_obsk is None:
+            return actions[self.possible_agents[0]]
+
+        assert self.gym_env.action_space.shape is not None
+        global_action = numpy.zeros((self.gym_env.action_space.shape[0],)) + numpy.nan
+        for agent_id, partition in enumerate(self.agent_action_partitions):
+            for act_index, body_part in enumerate(partition):
+                assert numpy.isnan(
+                    global_action[body_part.act_ids]
+                ), "FATAL: At least one gym_env action is doubly defined!"
+                global_action[body_part.act_ids] = actions[
+                    self.possible_agents[agent_id]
+                ][act_index]
+
+        assert not numpy.isnan(
+            global_action
+        ).any(), "FATAL: At least one gym_env action is undefined!"
+        return global_action
+
+    def map_global_action_to_local_actions(
+        self, action: numpy.ndarray
+    ) -> dict[str, numpy.ndarray]:
+        """
+        Args:
+            action:
+                An array representing the actions of the single agent for this domain
+        Returns:
+            A dictionary of actions to be performed by each agent
+        Raises:
+            AssertionError:
+                If the Agent action factorization sizes are badly defined
+        """
+        if self.agent_obsk is None:
+            return {self.possible_agents[0]: action}
+
+        local_actions = {}
+        for agent_id, partition in enumerate(self.agent_action_partitions):
+            local_actions[self.possible_agents[agent_id]] = numpy.array(
+                [action[node.act_ids] for node in partition]
+            )
+
+        # assert sizes
+        assert len(local_actions) == len(self.action_spaces)
+        for agent in self.possible_agents:
+            assert len(local_actions[agent]) == self.action_spaces[agent].shape[0]
+
+        return local_actions
+
+    def map_global_state_to_local_observations(
+        self, global_state: numpy.ndarray
+    ) -> dict[str, numpy.ndarray]:
+        """
+        Args:
+            global_state:
+                the global_state (generated from MaMuJoCo.state())
+        Returns:
+            A dictionary of states that would be observed by each agent given the 'global_state'
+        """
+        if self.agent_obsk is None:
+            return {self.possible_agents[0]: global_state}
+
+        class data_struct:
```

Comment:
I would move this outside the `MaMujoco` class to be a global class

Reply:
Why? This is a function specific data struct,

Reply:
Ok, looking at the function in more detail. Why does the class exist? Can we not have the class attributes as normally variables?

Reply:
No we, can not pass all 122 arguments normally ;-)

`obsk.build_obs` is written this way, to be extensible, it supports all possible `mujoco` environments this way (and potentially be able to support future `brax` based environments with some modifications)